File: expression_pipeline.py
# -------------------------------------
# 1. Import Libraries
# -------------------------------------
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from statsmodels.stats.multitest import multipletests

# -------------------------------------
# 2. Load Expression Data and Annotations
# -------------------------------------
# Load expression data
expr_data = pd.read_csv("/datos/rnaseq_demos/GSE67333_norm_counts_FPKM_GRCh38.p13_NCBI.tsv.gz", sep="\t")
expr_data.set_index("GeneID", inplace=True)

# Load sample annotations
annotations = pd.read_csv("/datos/rnaseq_demos/gse67333_labels.txt")
annotations["Condition"] = annotations["Condition"].str[:-1]  # Clean condition labels

# Reorder columns based on annotations
expr_data = expr_data[annotations["SampleID"]]

# -------------------------------------
# 3. Heatmap of Top Variable Genes
# -------------------------------------
# Calculate variance and select top 50 genes
expr_data["Variance"] = expr_data.var(axis=1)
top_genes = expr_data.sort_values("Variance", ascending=False).head(50).drop(columns=["Variance"])

# Plot heatmap
sns.heatmap(top_genes, cmap="viridis", xticklabels=True, yticklabels=False)
plt.title("Heatmap of Top 50 Variable Genes")
plt.show()

# -------------------------------------
# 4. PCA Analysis
# -------------------------------------
# Perform PCA on transposed data (samples as rows)
pca = PCA(n_components=2)
pca_results = pca.fit_transform(expr_data.T)

# Create a dataframe for PCA results
pca_df = pd.DataFrame(pca_results, columns=["PC1", "PC2"], index=expr_data.columns)
pca_df = pca_df.merge(annotations, left_index=True, right_on="SampleID")

# Scatterplot of PCA
sns.scatterplot(data=pca_df, x="PC1", y="PC2", hue="Condition", style="Condition", s=100)
plt.title("PCA of Samples")
plt.xlabel(f"PC1 ({pca.explained_variance_ratio_[0]*100:.1f}% variance)")
plt.ylabel(f"PC2 ({pca.explained_variance_ratio_[1]*100:.1f}% variance)")
plt.legend(bbox_to_anchor=(1, 1))
plt.show()

# -------------------------------------
# 5. Differential Expression Analysis (Simplified Limma-like Approach)
# -------------------------------------
# Define groups
group_ctrl = annotations.loc[annotations["Condition"] == "CTRL", "SampleID"]
group_load = annotations.loc[annotations["Condition"] == "LOAD", "SampleID"]

# Calculate log2 fold change
expr_data["logFC"] = np.log2(expr_data[group_load].mean(axis=1) + 1) - np.log2(expr_data[group_ctrl].mean(axis=1) + 1)

# Perform t-tests (as a proxy for differential expression)
from scipy.stats import ttest_ind

expr_data["P.Value"] = expr_data.apply(
    lambda row: ttest_ind(row[group_load], row[group_ctrl], equal_var=False).pvalue, axis=1
)

# Adjust p-values using Benjamini-Hochberg correction
expr_data["adj.P.Val"] = multipletests(expr_data["P.Value"], method="fdr_bh")[1]

# Save filtered results
filtered = expr_data[(expr_data["logFC"].abs() > 1) & (expr_data["adj.P.Val"] < 0.05)]
filtered.to_csv("/datos/rnaseq_demos/diffex_demo_python.txt", sep="\t")

# -------------------------------------
# 6. Functional Enrichment Analysis (Gene Ontology)
# -------------------------------------
# Load annotation for mapping
annot = pd.read_csv("/datos/rnaseq_demos/Human.GRCh38.p13.annot.tsv.gz", sep="\t")

# Map significant genes to HGNC symbols
diffex_with_hgnc = filtered.reset_index().merge(annot, left_on="GeneID", right_on="GeneID")
gene_list = diffex_with_hgnc["Symbol"].dropna().unique()

# Perform GO enrichment using gseapy
from gseapy import enrichr

# Enrichment analysis
enrich_results = enrichr(gene_list=gene_list, gene_sets="GO_Biological_Process_2021", organism="Human")

# Extract top results
top_go = enrich_results.results.sort_values("Adjusted P-value").head(10)

# Plot top results
sns.barplot(data=top_go, x="-log10(Adjusted P-value)", y="Term")
plt.title("Top GO Enriched Terms")
plt.xlabel("-Log10 Adjusted P-value")
plt.ylabel("GO Terms")
plt.show()

File: expression_pipeline.jl
# -------------------------------------
# 1. Load Required Libraries
# -------------------------------------
using DataFrames
using CSV
using StatsBase
using Plots

# -------------------------------------
# 2. Load Data
# -------------------------------------
# Load expression data
expr_data = CSV.read("/datos/rnaseq_demos/GSE67333_norm_counts_FPKM_GRCh38.p13_NCBI.tsv.gz", DataFrame)
rename!(expr_data, :GeneID => :gene_id)  # Rename the first column

# Load sample annotations
annotations = CSV.read("/datos/rnaseq_demos/gse67333_labels.txt", DataFrame)
annotations.Condition = replace.(annotations.Condition, r"(\d+)" => "")

# Ensure columns match annotations
expr_data = select(expr_data, [:gene_id, annotations.SampleID...])

# -------------------------------------
# 3. Heatmap of Top Variable Genes
# -------------------------------------
# Calculate variance for each gene
expr_data.variance = [var(row[2:end]) for row in eachrow(expr_data)]

# Select top 50 variable genes
top_genes = expr_data |> sort(:variance, rev = true) |> first(50)

# Plot heatmap
heatmap(Matrix(select(top_genes, Not([:gene_id, :variance]))),
    title = "Top 50 Variable Genes Heatmap",
    xlabel = "Samples",
    ylabel = "Genes",
    color=:viridis)

# -------------------------------------
# 4. PCA Analysis
# -------------------------------------
# Transpose expression data for PCA
sample_matrix = Matrix(select(expr_data, Not([:gene_id, :variance])))'
gene_ids = expr_data.gene_id

# Perform PCA
using MultivariateStats
pca_model = fit(PCA, sample_matrix, maxoutdim = 2)
pca_scores = projection(pca_model)

# Merge PCA results with annotations
pca_df = DataFrame(PC1 = pca_scores[:, 1], PC2 = pca_scores[:, 2], SampleID = annotations.SampleID, Condition = annotations.Condition)

# Plot PCA scatterplot
scatter(pca_df.PC1, pca_df.PC2,
    group=pca_df.Condition,
    title="PCA of Samples",
    xlabel="PC1",
    ylabel="PC2",
    label="Condition",
    legend=:topright)

# -------------------------------------
# 5. Differential Expression Analysis
# -------------------------------------
# Define groups
group_ctrl = findall(x -> x == "CTRL", annotations.Condition)
group_load = findall(x -> x == "LOAD", annotations.Condition)

# Compute log2 fold change
logFC = [mean(log2.(expr_data[group_load, i] .+ 1)) - mean(log2.(expr_data[group_ctrl, i] .+ 1)) for i in 2:size(expr_data, 2)]

# Perform t-tests
using HypothesisTests
p_values = [pvalue(TTest(log2.(expr_data[group_load, i] .+ 1), log2.(expr_data[group_ctrl, i] .+ 1))) for i in 2:size(expr_data, 2)]

# Adjust p-values (Benjamini-Hochberg)
using FDR
adj_p_values = FDR.adjust(p_values)

# Create results DataFrame
diffex_results = DataFrame(
    gene_id = expr_data.gene_id,
    logFC = logFC,
    PValue = p_values,
    adjPValue = adj_p_values
)

# Filter significant genes
filtered = filter(row -> abs(row.logFC) > 1 && row.adjPValue < 0.05, diffex_results)

# Save filtered results
CSV.write("/datos/rnaseq_demos/diffex_demo_julia.txt", filtered)

# -------------------------------------
# 6. Functional Enrichment Analysis
# -------------------------------------
# Map genes to HGNC symbols (assumes annotation file)
annot = CSV.read("/datos/rnaseq_demos/Human.GRCh38.p13.annot.tsv.gz", DataFrame)

# Merge significant genes with annotations
filtered_annot = leftjoin(filtered, annot, on = :gene_id)

# Perform a basic overrepresentation test (hypergeometric test)
function hypergeometric_enrichment(gene_set, background_set, all_genes)
    N = length(all_genes)  # Total genes
    K = length(background_set)  # Background size
    n = length(gene_set)  # Observed size
    x = length(intersect(gene_set, background_set))  # Intersection size
    Hypergeometric(K, N - K, n)(x)
end

# Example GO background (replace with real data)
example_go_set = ["GeneA", "GeneB", "GeneC", "GeneD"]
all_genes = annot.Symbol

# Run enrichment
enrichment_results = hypergeometric_enrichment(filtered_annot.Symbol, example_go_set, all_genes)

# -------------------------------------
# 7. Visualize Enrichment Results
# -------------------------------------
# (Here we mock enrichment visualization, adjust as needed)
bar(["GO Term 1", "GO Term 2"], [0.01, 0.05],
    title="Enriched GO Terms",
    xlabel="GO Terms",
    ylabel="-log10(P-Value)")
